### ISSUE TITLE
Making IfcConvert faster?

### ISSUE THREAD

**Issue by ostwalprasad** [IfcOpenShell/IfcOpenShell#1508]
Hello IfcOpenShell Community!

I'm converting the IFC file using IfcConvert to 4 different formats: XML, SVG, GLB, DAE.

While it works quite fast for smaller files <20MB, it takes really long time for larger files around 100MB. 

I'm using the python subprocess module to execute the commands.

Earlier, I was converting the four files one by one using `subprocess.run` like below, but it only utilizes a single-core CPU:
``` python
subprocess.run(["ifcconvert", "input.ifc", "output.glb", "-qy"]) 
subprocess.run(["ifcconvert", "input.ifc", "output.xml", "-qy"]) 
subprocess.run(["ifcconvert", "input.ifc", "output.svg", "-qy"]) 
subprocess.run(["ifcconvert", "input.ifc", "output.dae", "-qy"]) 
```

So now I'm running four parallel processes using python `subprocess.Popen`
For example, this is what I'm doing right now, so four different processes are spawned and work in parallel.
``` python
subprocess.Popen(["ifcconvert", "input.ifc", "output.glb", "-qy"]) 
subprocess.Popen(["ifcconvert", "input.ifc", "output.xml", "-qy"]) 
subprocess.Popen(["ifcconvert", "input.ifc", "output.svg", "-qy"]) 
subprocess.Popen(["ifcconvert", "input.ifc", "output.dae", "-qy"]) 
```
Here's the screenshot of `top` when the above command lines are executed in parallel 

![image](https://user-images.githubusercontent.com/18458949/120484985-3f267c00-c3d1-11eb-82ce-2136284488a7.png)

Executing conversion in parallel speeds up to some extent, but it is still slow. I saw that there's an option of using the thread flag in IfcConvert which uses multithreading, but there were few issues that mentioned it didn't convert geometry well in few cases, so I'm reluctant to use that right now.

I'm using general-purpose AWS EC2 T3 4-core Ubuntu instances right now to run the conversions. There are also Compute Optimized instances available which I'll try once.

Are there any suggested approaches to process large files parallelly? Also, any other ways to speed up the conversions for large files?

Thanks!

Edit: Added screenshot

**Comment by aothms** [comment]
> but there were few issues that mentioned it didn't convert geometry well in few cases

I think these days it should be relatively safe.

But yeah there are two main problems. We use OpenCASCADE as the geometry library which is simply not very fast when it concerns explicit meshes. For IFC4 tesselated items we very recently did some work to at least get it to process as fast as the 2x3 faceted polyhedra. You might benefit from that depending on the type of files you encounter.

The other problem maybe is that the interface to IfcConvert is not as versatile as should be and therefore, parsing and geometry processing has to be repeated if you need multiple outputs.

Both items are being worked on. The 1st is rather far-reaching as it requires a lot of architectural changes. The 2nd might flow out of some general work to also make the IfcConvert and serialization logic accessible from Python.

Lastly, if you don't trust IfcConvert's multi-threading you can do it manually by splitting up the IFC file yourself in Python. There is some ExtractElements recipies in here that can accomplish that rather easily.

**Comment by ostwalprasad** [comment]
Thanks for the reply.
Could you link the reference/document for splitting the IFC files?

**Comment by Moult** [comment]
Linking this to #4915. This is now basically solved but I'd like to wait until CGAL-Hybrid is battle tested a bit more and I can make it the default.

**Comment by Moult** [comment]
Finally resolved :) See #4915.
